subsys/bluetooth/controller/ll_sw/isoal.c

```diff
@@ -0,0 +1,414 @@
+
+		err |= err_alloc;
+
+		/*
+		 * For this SDU we can only consume of packet, bounded by:
+		 *   - What can fit in the destination SDU.
+		 *   - What remains of the packet.
+		 */
+		const size_t consume_len = MIN(
+			packet_available,
+			sink->sdu_production.sdu_available
+		);
+
+		LL_ASSERT(sdu_buf);
+		LL_ASSERT(pdu_payload);
+		if (pckt_meta->meta->status == ISOAL_PACKET_STATUS_VALID) {
+			memcpy(sdu_buf, pdu_payload + pckt_meta->meta->read_offset, consume_len);
+		}
+		sdu_buf     += consume_len;
+		pdu_payload += consume_len;
+		sink->sdu_production.sdu_written   += consume_len;
+		sink->sdu_production.sdu_available -= consume_len;
+		packet_available                   -= consume_len;
+
+		sdu->status |= pckt_meta->meta->status;
+
+		bool force_emit = (packet_available == 0) && is_end_fragment;
+
+		const isoal_status_t err_emit = isoal_rx_try_emit_sdu(sink, force_emit);
+
+		err |= err_emit;
+	}
+
+	LL_ASSERT(packet_available == 0);
+	return err;
+}
+
+
+/**
+ * @brief Consume a PDU: Copy contents into SDU(s) and emit to a sink
+ * @details Destination sink may have an already partially built SDU
+ *
+ * @param sink[in,out]    Destination sink with bookkeeping state
+ * @param pckt_meta[out]  Packet with meta information (origin, timing, status)
+ *
+ * @return TODO
+ */
+static isoal_status_t isoal_rx_packet_consume(struct isoal_sink *sink,
+					      const struct isoal_packet_rx *pckt_meta)
+{
+	isoal_packet_cnt_t id, prev_id;
+
+	isoal_status_t err = ISOAL_STATUS_OK;
+
+	/*
+	 *   SDU of length N can be partially filled, over a number of cases:
+	 *      0                          N-1
+	 *   1) [#########                   ] : We have the beginning. Not yet completed.
+	 *   2) [#########                   ] : We lack the end. Completed. Frame is lost.
+	 *   3) [                   #########] : We lack the beginning. Frame lost
+	 *   4) [         #########          ] : We lack the beginning. Not yet completed, sdu lost
+	 *   5) [#########          #########] : We lack something. Completed. Frame is lost.
+	 *   6) [                            ] : Predict
+	 */
+
+	const uint8_t llid = isoal_rx_packet_classify(pckt_meta);
+
+	/* TBD Below case statement is very redundant, might simplify when
+	 * development has progressed further
+	 */
+	switch (sink->sdu_production.fsm) {
+
+	case ISOAL_START:
+		/* State assumes First PDU of new SDU */
+		sink->sdu_production.prev_pdu_id = pckt_meta->meta->id;
+		if (llid == ISOAL_PACKET_START_OR_CONTINUATION) {
+			/* PDU contains first fragment of an SDU */
+			err |= isoal_rx_append_to_sdu(sink, pckt_meta, false);
+			sink->sdu_production.fsm = ISOAL_CONTINUE;
+		} else if (llid == ISOAL_PACKET_END_OR_COMPLETE) {
+			/* PDU contains complete SDU */
+			err |= isoal_rx_append_to_sdu(sink, pckt_meta, true);
+			sink->sdu_production.fsm = ISOAL_START;
+		} else {
+			/* Unsupported case */
+
+			/* TBD error handling */
```

Comment:
Maybe assert instead of silently ignoring the problem?

Reply:
Done